skrf/network.py

```diff
@@ -1109,62 +1112,48 @@ def z0(self):
-                    if len(self._z0) == self.frequency.npoints:
-                        # this z0 is frequency dependent but not port dependent
-                        self._z0 = \
-                            npy.repeat(npy.reshape(self._z0, (-1, 1)), self.number_of_ports, 1)
-
-                    elif len(self._z0) == self.number_of_ports:
-                        # this z0 is port dependent but not frequency dependent
-                        self._z0 = self._z0 * npy.ones( \
-                            (self.frequency.npoints, self.number_of_ports))
-
-                    else:
-                        raise (IndexError('z0 has bad shape'))
-
-                except AttributeError:
-                    # there is no self.frequency, or self.number_of_ports
-                    raise AttributeError('Error: I cant reshape z0 through inspection. you must provide correctly '
-                                         'shaped z0, or s-matrix first.')
-
+        if not hasattr(self, '_s'):
+            return self._z0
+        
+        if self._z0.ndim == 0:
+            return npy.full(self._s.shape[:2], self._z0)
+        elif self._z0.ndim == 1:
+            z0 = npy.ones(self._s.shape[:2], dtype=complex)
+            if len(self._z0) == self.nports:
+                z0[:] = self._z0[None, :]
+            #elif len(self._z0) == self.frequency.npoints:
+            else:
+                z0[:] = self._z0[:,None]
+            return z0
+        elif self._z0.ndim == 2:
             return self._z0
-
-        except AttributeError:
-            # print('Warning: z0 is undefined. Defaulting to 50.')
-            self.z0 = 50
-            return self.z0  # this is not an error, its a recursive call
 
     @z0.setter
     def z0(self, z0):
-        """z0=npy.array(z0)
-        if len(z0.shape) < 2:
-                try:
-                        #try and re-shape z0 to match s
-                        z0=z0*npy.ones(self.s.shape[:-1])
-                except(AttributeError):
-                        print ('Warning: you should store a Network\'s \'s\' matrix before its \'z0\'')
-                        #they have yet to set s .
-                        pass
-        """
-        self._z0 = npy.array(z0, dtype=complex)
+        if isinstance(z0, Number):
+            self._z0 = npy.array(z0, dtype=complex)
+            return
+
+        z0 = npy.array(z0, dtype=complex)
+
+        if z0.ndim == 0:
+            self._z0 = z0
+            return
+        if not hasattr(self, '_s'):
+            if 1 <= z0.ndim <= 2:
+                self._z0 = z0
+                return
+
+        if z0.ndim == 1:
+            if len(z0) in (self.frequency.npoints, self.nports):
+                self._z0 = z0
+                return
+        elif z0.ndim == 2:
+            if z0.shape == (self.frequency.npoints, self.nports):
+                self._z0 = z0
+                return
+
+        raise AttributeError('Unable to broadcast z0 to s shape')
```

Comment:
Ok. would be nice if you could add some comments to explain the different cases

Reply:
i agree that a few comments would go a  long way. looks great though. 
adding tests for the various z0  setters would be great since you have it fresh in mind, but up to you. 

thanks.

Reply:
Thanks for the feedback, I added some comments in both setter and gettter.

I will add some tests the next days :)